app/Jobs/FetchNewJobs.php

```diff
@@ -61,9 +61,20 @@ public function handle(): void
         }
 
         // Fire the event
-        // @TODO: Apply filters here
-        $jobs = JobPost::visible()->unnotified()->orderBy('published_at', 'desc')->get()->all();
-        event(new JobsPosted($jobs, $this->notifyEmpty));
+        $this->notifyJobsPosted();
+    }
+
+    /**
+     * Notify the user of new jobs that have been posted.
+     * If the user has defined filters for their notification preference,
+     * we'll only notify them of jobs that match their criteria.
+     *
+     * The event constructor is responsible for fetching the
+     * jobs which meet the criteria.
```

Comment:
Why is the event constructor responsible for gathering the data? It seems backwards to me, you wouldn't be able to reuse the event for anything else taking that approach.

What are the benefits of making the event prepare its own data instead of passing the relevant data to the event when creating it? I'm not sure I've seen that pattern before so I'm genuinely curious.

Reply:
It's a good observation - I mainly did this to increase testability. I wanted our test to be able to prove that only jobs meeting the filter criteria are notified.

If we're passing the jobs to the event, we can't reliably test _which_ jobs the user will be notified of when `FetchNewJobs` is run, unless we were to copy over the logic of the query into the test (but then we'd end up with a brittle test)

I think we've got three options:

- Make the event responsible for fetching the jobs. (The change made in this PR). This way, both our tests and the real implementation will be using the same logic to fetch jobs to notify the user about. As you've mentioned, the downside of this is that the event is only usable for notifying the user of new jobs that meet their criteria and haven't already been notified of yet.
- Make the listener responsible for fetching the jobs. This would give us the same result as the above option, but at least we could re-use the event by attaching different listeners.
- Refactor the FetchNewJobs job so that it has a service responsible for fetching the data from the RSS feed. That way, we can mock the response of the RSS feed in our tests. I think that's a good change to make anyway because it'll allow us to test the persistence part of this job, but it still doesn't really give us a way to test the notification sending part _in isolation_.

What do you think? I'm leaning towards option three. We'll make the `JobsPostedEventTest` responsible for testing that notifications are sent (not _which_ jobs the user is notified of), and we'll make a new test for `FetchNewJobs` which is responsible for testing that the jobs are persisted correctly _and_ the event is dispatched with only jobs the user is interested in.

**TLDR**

- did it for testability
- thought the event should be responsible for determining if the user wants the notification.
- now I've changed my mind, I'll refactor and make a new test for testing that bit